examples/forest_fire.jl

```diff
@@ -16,116 +16,113 @@
 
 # ## Defining the core structures
 
-# We start by defining the agent type
+# Cellular automata don't necessarily require an agent-like structure. Here we will
+# demonstrate how a model focused solution is possible.
 using Agents, Random
-using InteractiveDynamics
 using CairoMakie
 
-mutable struct Tree <: AbstractAgent
+mutable struct Automata <: AbstractAgent
     id::Int
-    pos::Dims{2}
-    status::Symbol  #:green, :burning, :burnt
 end
 nothing # hide
 
-# The agent type `Tree` has three fields: `id` and `pos`, which have to be there for any agent,
-# and a `status` field that we introduce for this specific model.
-# The `status` field will be `:green` when the tree is ok, `:burning` when on fire,
-# and finally `:burnt`.
+# The agent type `Automata` is effectively a dummy agent, for which we will invoke
+# [`dummystep`](@ref) when stepping the model.
 
 # We then make a setup function that initializes the model.
 function forest_fire(; density = 0.7, griddims = (100, 100))
-    space = GridSpace(griddims; periodic = false, metric = :euclidean)
-    forest = AgentBasedModel(Tree, space)
-    ## create and add trees to each position with a probability
-    ## determined by the `density`.
-    for position in positions(forest)
+    offsets = [Tuple(i == j ? 1 : 0 for i in 1:2) for j in 1:2]
+    offsets = vcat(offsets, [.-dir for dir in offsets])
+    ## The `trees` field is coded such that
+    ## Empty = 0, Green = 1, Burning = 2, Burnt = 3
+    ## `offsets` is a quick way of identifying euclidean neighbors.
+    forest = ABM(Automata; properties = (trees = zeros(Int, griddims), offsets = offsets))
+    for I in CartesianIndices(forest.trees)
         if rand(forest.rng) < density
-            ## Set the trees at position x=1 on fire
-            state = position[1] == 1 ? :burning : :green
-            add_agent!(position, forest, state)
+            ## Set the trees at the left edge on fire
+            forest.trees[I] = I[1] == 1 ? 2 : 1
         end
     end
     return forest
 end
 
+# Notice we have not even required the use of a space for this simple model.
+
 forest = forest_fire()
 
 # ## Defining the step!
-# Because of the way the forest fire model is defined, we only need a
-# stepping function for the agents
-
-function tree_step!(tree, forest)
-    ## The current tree is burning
-    if tree.status == :burning
-        ## Find all green neighbors and set them on fire
-        for neighbor in nearby_agents(tree, forest)
-            if neighbor.status == :green
-                neighbor.status = :burning
+
+function tree_step!(forest)
+    nx, ny = size(forest.trees)
+    ## Find trees that are burning (coded as 2)
+    for I in findall(isequal(2), forest.trees)
+        ## Look up all euclidean neighbors
+        neighbors = Iterators.filter(
+            x -> 1 <= x[1] <= nx && 1 <= x[2] <= ny,
+            (I.I .+ n for n in forest.offsets),
+        )

```

Comment:
Hm, I'm a bit afraid with this one. Is it that much slower to use `nearby_positions`? Because someone could complain "Well, you're not really using Agents.jl anymore but writing the implementation from scratch, including even finding neighboring positions".

Reply:
To do that we'd also add a space that does nothing. I can check.

Reply:
aaaaaaaaaaaaaaaaaaaa aaaaaaa sorry yeah you are right...

I wonder if we can somehow re-use the `Hood` stuff. Hm... but maybe not worth the development ATM.

Reply:
A no, what am I even talking about. We can add the space and just use `neaby_positions` to get the indices with which to access the `forest.tree` array. We don't have to actually develop anything as long as the space and the `trees` have the same size.